libs/go/libp2p_node/aea/api.go

```diff
@@ -397,23 +427,97 @@ func (aea *AeaApi) stop() {
   Pipes helpers
 
 */
+const CurrentVersion = "0.1.0"
 
-func write_envelope(pipe Pipe, envelope *Envelope) error {
-	data, err := proto.Marshal(envelope)
+func MakeACNMessageFromEnvelope(envelope *Envelope) (error, []byte) {
+	envelope_bytes, err := proto.Marshal(envelope)
+	if err != nil {
+		return err, envelope_bytes
+	}
+	return acn.EncodeACNEnvelope(envelope_bytes)
+}
+
+func (aea AeaApi) SendEnvelope(envelope *Envelope) error {
+	err, data := MakeACNMessageFromEnvelope(envelope)
 	if err != nil {
 		logger.Error().Str("err", err.Error()).Msgf("while serializing envelope: %s", envelope)
 		return err
 	}
-	return pipe.Write(data)
+	err = aea.pipe.Write(data)
+	if err != nil {
+		logger.Error().Str("err", err.Error()).Msgf("on pipe write")
+		return err
+	}
+
+	status, err := acn.WaitForStatus(aea.acn_status_chan, ACN_STATUS_TIMEOUT)
+
+	if err != nil {
+		logger.Error().Str("err", err.Error()).Msgf("on status wait")
+		return err
+	}
+	if status.Code != acn.Status_SUCCESS {
+		logger.Error().Str("err", err.Error()).Msgf("bad status: %d", status.Code)
+		return errors.New("bad status!")
+	}
+	return err
 }
 
-func read_envelope(pipe Pipe) (*Envelope, error) {
+func (aea AeaApi) AddACNStatusMessage(status *acn.Status, counterpartyID string) {
+	aea.acn_status_chan <- status
+	logger.Info().Msgf("chan len is %d", len(aea.acn_status_chan))
+}
+
+func (aea AeaApi) ReceiveEnvelope() (*Envelope, error) {
 	envelope := &Envelope{}
-	data, err := pipe.Read()
+	var acn_err error
+	data, err := aea.pipe.Read()
+
 	if err != nil {
 		logger.Error().Str("err", err.Error()).Msg("while receiving data")
 		return envelope, err
 	}
-	err = proto.Unmarshal(data, envelope)
-	return envelope, err
+	msg_type, acn_envelope, status, err := acn.DecodeACNMessage(data)
+
+	if err != nil {
+		logger.Error().Str("err", err.Error()).Msg("while decoding acn")
+		acn_err = acn.SendAcnError(aea.pipe, "error on decoding acn message")
+		if acn_err != nil {
+			logger.Error().Str("err", err.Error()).Msg("on acn send error")
+		}
+		return envelope, err
+	}
+
+	switch msg_type {
+	case "aea_envelope":
+		{
+			err = proto.Unmarshal(acn_envelope.Envel, envelope)
+			if err != nil {
+				logger.Error().Str("err", err.Error()).Msg("while decoding envelope")
+				acn_err = acn.SendAcnError(aea.pipe, "error on decoding envelope")
+				if acn_err != nil {
+					logger.Error().Str("err", err.Error()).Msg("on acn send error")
+				}
+				return envelope, err
+			}
+			err = acn.SendAcnSuccess(aea.pipe)
+			return envelope, err
+
+		}
+	case "status":
+		{
+			logger.Info().Msgf("acn status %d", status.Code)
+			aea.acn_status_chan <- status
```

Comment:
Note; since both ReceiveEnvelope and SendEnvelope are executed in different co-routines this can lead to a situation where more than one status message needs to be put on the channel ...